wrapper/src/main/java/software/amazon/jdbc/plugin/ReadOnlyEndpointConnectionPlugin.java

```diff
@@ -0,0 +1,140 @@
+          add("connect");
+          add("forceConnect");
+        }
+      });
+
+  private final PluginService pluginService;
+  private HostListProviderService hostListProviderService;
+  private final RdsUtils rdsHelper;
+
+  public ReadOnlyEndpointConnectionPlugin(final PluginService pluginService) {
+    this(pluginService, new RdsUtils());
+  }
+
+  private ReadOnlyEndpointConnectionPlugin(final PluginService pluginService, final RdsUtils rdsHelper) {
+    this.pluginService = pluginService;
+    this.rdsHelper = rdsHelper;
+  }
+
+  @Override
+  public Set<String> getSubscribedMethods() {
+    return subscribedMethods;
+  }
+
+  @Override
+  public void initHostProvider(
+      final String driverProtocol,
+      final String initialUrl,
+      final Properties props,
+      final HostListProviderService hostListProviderService,
+      final JdbcCallable<Void, SQLException> initHostProviderFunc)
+      throws SQLException {
+
+    this.hostListProviderService = hostListProviderService;
+    initHostProviderFunc.call();
+  }
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, connectFunc);
+  }
+
+  private Connection connectInternal(
+      HostSpec hostSpec,
+      Properties props,
+      JdbcCallable<Connection, SQLException> connectFunc) {
+    try {
+      final Connection currentConnection = connectFunc.call();
+
+      if (currentConnection != null && this.rdsHelper.isReaderClusterDns(hostSpec.getHost())) {
+        final Dialect dialect = this.pluginService.getDialect();
+        String nodeIdQuery = "";
+
+        if (dialect instanceof AuroraDialect) {
+          final AuroraDialect auroraDialect = (AuroraDialect) dialect;
+          nodeIdQuery = auroraDialect.getNodeIdQuery();
+        }
+
+        if (!StringUtils.isNullOrEmpty(nodeIdQuery)) {
+          final String connectedHostName = getCurrentlyConnectedInstance(currentConnection, nodeIdQuery);
+
+          // Replace the connection created using the read only endpoint with an instance endpoint using the queried
+          // instance name
+          if (!StringUtils.isNullOrEmpty(connectedHostName)) {
+            final String pattern = props.getProperty(CLUSTER_INSTANCE_HOST_PATTERN);
+            String instanceEndpoint = !StringUtils.isNullOrEmpty(pattern) ? pattern :
+                this.rdsHelper.getRdsInstanceHostPattern(hostSpec.getHost());
+            instanceEndpoint = instanceEndpoint.replace("?", connectedHostName);
+            final HostSpec newHostSpec = this.pluginService.getHostSpecBuilder().host(instanceEndpoint).build();
+            this.hostListProviderService.setInitialConnectionHostSpec(newHostSpec);
+            return this.pluginService.connect(newHostSpec, props);
```

Comment:
If the intent is to return the instance endpoint of the reader we are currently connected to using the readonly cluster endpoint, why are we calling connect again? What about the connection we have already established?